2014-10-01  Magnus Granberg  <zorry@gentoo.org>

	We copy Scheduler.py from portage and patch it.
	Fix so we can use add_buildlog_main()
	We use add_buildlog_main() for loging.

--- a/zobcs/pym/Scheduler.py	2013-03-22 17:57:23.000000000 +0100
+++ b/zobcs/pym/Scheduler.py	2012-12-21 02:09:28.082301168 +0100
@@ -62,6 +62,8 @@ from _emerge.PackageMerge import Package
 from _emerge.PollScheduler import PollScheduler
 from _emerge.SequentialTaskQueue import SequentialTaskQueue
 
+from zobcs.build_log import add_buildlog_main
+
 if sys.hexversion >= 0x3000000:
 	basestring = str
 
@@ -1254,8 +1251,9 @@ class Scheduler(PollScheduler):
 
 	def _do_merge_exit(self, merge):
 		pkg = merge.merge.pkg
+		settings = merge.merge.settings
+		trees = self.trees
 		if merge.returncode != os.EX_OK:
-			settings = merge.merge.settings
 			build_dir = settings.get("PORTAGE_BUILDDIR")
 			build_log = settings.get("PORTAGE_LOG_FILE")
 
@@ -1266,6 +1264,7 @@ class Scheduler(PollScheduler):
 			if not self._terminated_tasks:
 				self._failed_pkg_msg(self._failed_pkgs[-1], "install", "to")
 				self._status_display.failed = len(self._failed_pkgs)
+			add_buildlog_main(settings, pkg, trees)
 			return
 
 		self._task_complete(pkg)
@@ -1284,6 +1283,7 @@ class Scheduler(PollScheduler):
 				self._pkg_cache.pop(pkg_to_replace, None)
 
 		if pkg.installed:
+			add_buildlog_main(settings, pkg, trees)
 			return
 
 		# Call mtimedb.commit() after each merge so that
@@ -1294,6 +1294,7 @@ class Scheduler(PollScheduler):
 		if not mtimedb["resume"]["mergelist"]:
 			del mtimedb["resume"]
 		mtimedb.commit()
+		add_buildlog_main(settings, pkg, trees)
 
 	def _build_exit(self, build):
 		self._running_tasks.pop(id(build), None)
@@ -1318,6 +1319,8 @@ class Scheduler(PollScheduler):
 				self._status_display.merges = len(self._task_queues.merge)
 		else:
 			settings = build.settings
+			trees = self.trees
+			pkg = build.pkg
 			build_dir = settings.get("PORTAGE_BUILDDIR")
 			build_log = settings.get("PORTAGE_LOG_FILE")
 
@@ -1329,6 +1332,7 @@ class Scheduler(PollScheduler):
 				self._failed_pkg_msg(self._failed_pkgs[-1], "emerge", "for")
 				self._status_display.failed = len(self._failed_pkgs)
 			self._deallocate_config(build.settings)
+			add_buildlog_main(settings, pkg, trees)
 		self._jobs -= 1
 		self._status_display.running = self._jobs
 		self._schedule()
